### ISSUE TITLE
Extremely slow and laggy performance comparing to default terminal

### ISSUE THREAD

**Issue by coder3119** [warpdotdev/Warp#4856]
### Describe the bug

 I am experiencing performance issues with the Warp terminal when running it on a remote server with the latest version of Kali Linux using the GNOME environment. The Warp terminal appears to be slow and laggy compared to a normal terminal in this setup.

### To reproduce

Steps to Reproduce:

Set up a remote server running the latest version of Kali Linux with the GNOME desktop environment installed.
Install and open the Warp terminal on the remote server.
Execute various commands within the Warp terminal and observe the response time and overall performance.
Compare the performance with the standard GNOME terminal by executing similar commands.

### Expected behavior

Expected Result: The Warp terminal should function with comparable efficiency and responsiveness as the standard GNOME terminal.

Actual Result: The Warp terminal exhibits slower response times and a general lack of smoothness in operation, significantly impacting usability.

Frequency: This issue occurs consistently every time the Warp terminal is used in the described environment.

### Screenshots

![warp](https://github.com/warpdotdev/Warp/assets/159404436/e8e101e0-d736-4259-a4f1-7515eaa94335)

### Operating system

Linux

### Operating system and version

Lastest Kali linux ( kernel Linux 6.6.9-amd64 )

### Shell Version

zsh 5.9 (x86_64-debian-linux-gnu)

### Current Warp version

v0.2024.04.23.08.01.stable_03

### Regression

Yes, this bug started recently or with an X Warp version

### Recent working Warp date

warp-terminal_0.2024.04.02.08.02.stable.01_amd64.deb

### Additional context

i had the version 0.2024.04.02.08.02.stable.01 on older similar setup and it was working fine.

### Does this block you from using Warp daily?

Yes, this issue prevents me from using Warp daily.

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by acarl005** [comment]
Hi @coder3119, thanks for reporting this case. Can you elaborate how you're connected to this remote server, e.g. is it VNC, RDP, X11 forwarding over SSH?

**Comment by acarl005** [comment]
Looking at #4864, it seems that this may not be limited to remote Linux sessions.

**Comment by coder3119** [comment]
I'm using VMware Remote Console (VMRC) but the same behavior when i use RDP.
when i switched to using Xfce, I noticed a slightly improvement

**Comment by alokedesai** [comment]
@coder3119 We believe we just pushed a fix for this in the latest version of Warp (`v0.2024.04.30.08.02.stable_02`). Would you mind updating and confirming it fixes things for you?

**Comment by elviskahoro** [comment]
Fixed a bug! Please ignore the devx bot.

**Comment by tothpsk** [comment]
I was experiencing the same issue on Debian Testing, Warp version v0.2024.09.03.08.02.stable_03. My OS was running on Wayland. However, changing to X11 solved the problem and Warp is launching pretty fast now :-)

**Comment by tothpsk** [comment]
Hi, thank you so much for addressing this issue in the latest release (v0.2024.09.17.08.02.stable_01) . I updated my installation, but it seems it is not fixed. It still takes ages to open Warp on Wayland. With X11 the app opens within 1-2 seconds.

**Comment by ifedan-ed** [comment]
After upgrading to Ubuntu 24.10, warp take an extremely amount of time to launch rendering it unusable

**Comment by alokedesai** [comment]
Hi @tothpsk and @ifedan-ed. Could you both provide your Warp logs + a video demonstrating the behavior so we can debug further? Thank you! https://docs.warp.dev/help/sending-us-feedback#gathering-warp-logs

**Comment by tothpsk** [comment]
Hi @alokedesai, here are the videos and logs:

launching Warp on X11: https://github.com/user-attachments/assets/3859e9bf-650f-451a-ba36-0ee28d848a92
launching Warp on Wayland: https://github.com/user-attachments/assets/80f45483-6296-44ab-a20a-77e040e7371e
logs: [warp-terminal.zip](https://github.com/user-attachments/files/17178794/warp-terminal.zip)

Hope this helps finding the cause of the issue. Best wishes,

Peter

**Comment by hasan4temp** [comment]
facing the same issue on v0.2024.12.10.15.55.stable_03 WSL 2

**Comment by TylerHoekstra** [comment]
facing the same issue on v0.2025.01.08.02.stable_04 running on WSL2. The device I'm on has no dedicated graphics card and the CPU is constantly getting maxed out by VmmemWSL (in this case warp terminal).

**Comment by AnouarTouati** [comment]
I am having the same issue running Wrap on my desktop Ubuntu 24.04.1 LTS. It does not feel snappy to type. (The video provided is a bit slower than normal cause I was recording.)

https://github.com/user-attachments/assets/ea7de3af-0a5c-4cd1-882b-8d2d15945f7f

**Comment by sewandev** [comment]
It's April 2025 and Warp Terminal is still extremely slow for Windows. :(

**Comment by alokedesai** [comment]
@sewandev Can you describe the performance issues you're seeing on the Windows? Would you be free to join a meeting with us to dig into the issues further? We'd be happy to give you a gift (warp swag or a gift card) as a thank you for your time.

**Comment by ori-scala** [comment]
@alokedesai 
```
zprof
num  calls                time                       self            name
-----------------------------------------------------------------------------------
 1)    1       31797.15 31797.15   57.49%  30254.19 30254.19   54.70%  warp_bootstrapped
 2)    1       19343.16 19343.16   34.97%  11220.63 11220.63   20.29%  warp_precmd
 3)    5       10442.50  2088.50   18.88%  10442.50  2088.50   18.88%  warp_send_json_message
```
It's becoming unusable

**Comment by GurdipS5** [comment]
slow for me on windows too.

**Comment by alokedesai** [comment]
@GurdipS5 What issues are you seeing on Windows? Would you mind filing a new issue with more details and tagging me on it?

**Comment by victorhverat** [comment]
Lento tambien en linux debian, consumo de prosesador y al escribir tarda en mostrar, hasta que se arregle continuo con alacritty

**Comment by alokedesai** [comment]
@Isolator79 What OS are you on? Could you send me a video of the exact behavior you're seeing when you login with Google vs email so I can debug further? Thanks!

My email is aloke@warp.dev

**Comment by itobey** [comment]
This will probably not help you, but for the record for everyone else stumbling across this issue: I've experienced this issue on Windows and disabling the integrated GPU usage (Warp Settings -> Features -> System -> Prefer rendering new windows with integrated GPU) fixed this issue for me.

**Comment by Jordan-Haidee** [comment]
> This will probably not help you, but for the record for everyone else stumbling across this issue: I've experienced this issue on Windows and disabling the integrated GPU usage (Warp Settings -> Features -> System -> Prefer rendering new windows with integrated GPU) fixed this issue for me.

Thanks! 👍

**Comment by ZarK** [comment]
Warp team, please look into this issue - Warp on Windows is still almost unusable from lag. I have tried all tips above and from Discord, but it does not help. 

As long as Warp is the foreground window, the whole Windows OS is laggy, and especially Warp is extremely laggy. If I focus another window, it takes a second and then the lag disappear. Focusing Warp again and the lag is immediately back.

What has been tried:

- Disable “Prefer rendering new windows with integrated GPU (low power)” in Settings > Features > System. This resolved input lag/stutter for multiple users. [[Input lag](https://github.com/warpdotdev/Warp/issues/7276); [Extremely slow](https://github.com/warpdotdev/Warp/issues/4856)].
- Switch Preferred graphics backend to DirectX 12 (then restart Warp). Some issues were fixed by moving off Vulkan. [[Performance issues](https://github.com/warpdotdev/Warp/issues/6038)]
- Force Warp to use the discrete GPU via Windows Graphics Settings or your GPU control panel. [[Warp opens once](https://github.com/warpdotdev/Warp/issues/6105); [Performance issues](https://github.com/warpdotdev/Warp/issues/6038)]
- On NVIDIA, set “Preferred refresh rate” to “Highest available” for warp.exe in NVIDIA Control Panel. This fixed heavy FPS drops for some. [[Poor performance](https://github.com/warpdotdev/Warp/issues/6564)]. I also tried setting Low Latency Mode to Ultra, Vertical Sync to Off and Power Management Mode to max.
- If Warp won’t honor backend selection or still lags, try setting WGPU_BACKEND=dx12 as a user env var, then restart. [[Not launching](https://github.com/warpdotdev/Warp/issues/6078)]. I also tried WGPU_POWER_PREF=high. Other WGPU apps have no issue on my system.
- Tested minimize Warp or move it to a different monitor—some users see lag only when Warp is focused or on a specific display. [[Poor performance](https://github.com/warpdotdev/Warp/issues/6564)]
- Tested setting CPU affinity to only 3D-cores or only non-3D cores on my 7950X3D CPU
- Tested changing refresh rate on displays from 144 to 60.

Windows 11, RTX 4090, Ryzen 7950X3D.

**Comment by codename-cn** [comment]
Still same problem on debian.

**Comment by GurdipS5** [comment]
This tool is wank

**Comment by SpartanTech** [comment]
Did anyone figure out fps / refresh rate / timer issues with Warp? Fresh install windows 11, decided to give this a try. anytime the window is up, the entire monitor lags. bad fps, mouse movement not smooth. all i have to do is minimize Warp and back to normal. fresh install of everything. Tried OpenGL, DirectX 12, The Nvidia exclusion, Vulcan, etc. Nothing. I think most people wouldnt notice if they aren't use to smooth or high refresh rates. Sad, was really looking forward to using this?

Edit: Did NOT Fix it :/ Only thought it did because I was clicked off Warp. Christiaan's exclusion allows the performance to perform correctly when Warp is NOT focused though, previously required an entire minimization. So I guess a step in the right direction.

**Comment by ChristiaanKl** [comment]
> Did anyone figure out fps / refresh rate / timer issues with Warp? Fresh install windows 11, decided to give this a try. anytime the window is up, the entire monitor lags. bad fps, mouse movement not smooth. all i have to do is minimize Warp and back to normal. fresh install of everything. Tried OpenGL, DirectX 12, The Nvidia exclusion, Vulcan, etc. Nothing. I think most people wouldnt notice if they aren't use to smooth or high refresh rates. Sad, was really looking forward to using this?
> 
> Edit: Did NOT Fix it :/ Only thought it did because I was clicked off Warp. Christiaan's exclusion allows the performance to perform correctly when Warp is NOT focused though, previously required an entire minimization. So I guess a step in the right direction.
> 
> Edit2: FIXED! (Workaround). Noticed it was only happening on CERTAIN monitors. moving it to another yieled NO performance loss (visible at least). This led me to think something else. FPGA module and DWM.
> 
> Simple and elegant: in **fullscreen-only** mode, G-Sync is completely dormant while DWM is compositing the desktop. The FPGA module hands control back to DWM which runs at my fixed 120Hz. Warp's chaotic frame submissions no longer perturb the panel's pixel clock because VRR isn't even active. The FPGA is just a passive fixed-refresh panel now until something takes exclusive fullscreen.
> 
> With Gsync on, any app presenting frames could influence the VRR cadence. Most apps are fine. Warp is not.
> 
> Make sure to restart Warp when you test this.
> 
> If I could see the code, I would look at In order of importance:
> 
> **1. Frame pacing — the main fix?** When no terminal content has changed, present with `sync_interval = 1` (wait for vsync) instead of submitting immediately. One line change in their wgpu present call. This alone would fix your issue entirely because the FPGA would receive frames at clean 120Hz intervals and maintain stable lock.
> 
> **2. Idle detection** Track whether any terminal output, cursor blink, or UI animation has occurred since the last frame. If nothing changed, don't submit a new frame at all, or submit at a very low rate like 10fps. From their code, this is how WezTerm, Alacritty, and Kitty all work. I think it may also fix that GitHub issue about excessive CPU/GPU use while idle, same root cause.
> 
> **3. Explicit swap chain flags** In their wgpu/DXGI initialization, declare `DXGI_SWAP_EFFECT_FLIP_DISCARD` with `DXGI_PRESENT_ALLOW_TEARING` only when tearing is actually desired. This opts them into the flip model on their own terms rather than having Windows' "Optimizations for Windowed Games" try to upgrade them automatically.
> 
> **4. Respect the VRR range** If they're going to render above vsync, cap themselves to the display's reported max VRR rate via DXGI's `GetDesc` , ex for my ultrawide that's 120Hz. Never submit faster than the panel can track.
> 
> The whole fix is probably 50–100 lines of code changes in their rendering initialization and main loop. It's not architecturally complex, it's just render loop hygiene that games have known to do for 15 years. The fact that it ships without this on Windows suggests their primary development target is still macOS/Linux and Windows is playing catch-up.
> 
> TLDR: **So workaround 1:** Move to another monitor. See if it has the same problem. **Workaround 2:** Disable GSync for that monitor (Or try only full screen mode). Requires Warp restart.
> 
> Interested to see if this does not solve anyone's issue.

Interesting... I still, to this day have not had any issues with Warp after the changes I made. And it doesn't surprise me, because more often than not, a fix for one person might not work for another. I never expected GSync to have a performance impact if I am being honest. Thanks for the update.

**Comment by ZarK** [comment]
@acarl005 this was unrelated to Intel and also experimented on AMD boards / chips with no Intel GPU.

**Comment by acarl005** [comment]
Thanks for clarifying @ZarK. The issue started to balloon its scope beyond any specific issue. I picked one specific cause to attribute this to and closed it out. I'll look into your issue next. In Warp can you open the command palette (ctrl-shift-P) and search for "dump debug info" and post the output of that command?

**Comment by acarl005** [comment]
For those still having performance issues on other hardware, please create a new issue for your hardware. Thanks!

**Comment by panicoenlaxbox** [comment]
For me, this is unusable — it's slow, and it's a shame because I really want to use it.

<img width="1466" height="320" alt="Image" src="https://github.com/user-attachments/assets/f97c92df-16fb-4e02-a1d8-58693deb644b" />

**Comment by acarl005** [comment]
@panicoenlaxbox In Warp can you open the command palette (ctrl-shift-P) and search for "dump debug info" and post the output of that command?

Alternatively you can do it from a different terminal from PowerShell

```
# If Warp is in your PATH
warp --dump-debug-info

# otherwise you may need to use an absolute path, depending on your installation...
& $env:LOCALAPPDATA\programs\Warp\warp.exe --dump-debug-info
& $env:PROGRAMFILES\Warp\warp.exe --dump-debug-info
```

**Comment by GurdipS5** [comment]
This is wank

**Comment by panicoenlaxbox** [comment]
Hi @acarl005,

Maybe 'unusable' was an exaggeration, but there's a delay (a few milliseconds) when typing, opening autocomplete, or selecting an option from that list (and it doesn't always happen) that ruins the experience. That delay doesn't show up in other applications (nor in Windows Terminal). Either way, the app doesn't feel fast — it's 'a little bit' laggy, but enough to break the flow

In fact, at first I enabled the option to expand aliases, and I had to remove it because then the issue was more noticeable

Thanks!

```
C:\Users\sergio.leon\AppData\Local\Programs\Warp\warp.exe --dump-debug-info
Warp version: Some("v0.2026.07.01.09.21.stable_01")
gpu_power_preference: HighPerformance
backend_preference: None
windowing_system: None
##################################################
# wgpu Adapters
##################################################
DiscreteGpu: NVIDIA GeForce RTX 4060 Laptop GPU
	Backend: Dx12
	Driver: 32.0.16.1047
IntegratedGpu: Intel(R) Arc(TM) Graphics
	Backend: Dx12
	Driver: 32.0.101.8132
IntegratedGpu: Intel(R) Arc(TM) Graphics
	Backend: Vulkan
	Driver: Intel Corporation (101.8132)
Other: NVIDIA GeForce RTX 4060 Laptop GPU/PCIe/SSE2
	Backend: Gl
	Driver: ? (4.6.0 NVIDIA 610.47)
Cpu: Microsoft Basic Render Driver
	Backend: Dx12
	Driver: 10.0.26100.8521
DiscreteGpu: NVIDIA GeForce RTX 4060 Laptop GPU
	Backend: Vulkan
	Driver: NVIDIA (610.47)
```
